bridge-history-api/internal/orm/cross_message.go

```diff
@@ -0,0 +1,395 @@
+			updateFields["l1_tx_hash"] = l1MessageQueueEvent.TxHash.String()
+			db = db.Where("message_type", MessageTypeL1SentMessage)
+			db = db.Where("message_nonce = ?", l1MessageQueueEvent.QueueIndex)
+			if err := db.Updates(updateFields).Error; err != nil {
+				return fmt.Errorf("failed to update L1 tx hash of QueueTransaction, event: %+v, error: %w", l1MessageQueueEvent, err)
+			}
+		}
+		if l1MessageQueueEvent.EventType == MessageQueueEventTypeDequeueTransaction {
+			updateFields := make(map[string]interface{})
+			updateFields["tx_status"] = TxStatusTypeSkipped
+			db = db.Where("message_type", MessageTypeL1SentMessage)
+			db = db.Where("message_nonce = ?", l1MessageQueueEvent.QueueIndex)
+			if err := db.Updates(updateFields).Error; err != nil {
+				return fmt.Errorf("failed to update message status as skipped, event: %+v, error: %w", l1MessageQueueEvent, err)
+			}
+		}
+		if l1MessageQueueEvent.EventType == MessageQueueEventTypeDropTransaction {
+			updateFields := make(map[string]interface{})
+			updateFields["tx_status"] = TxStatusTypeDropped
+			db = db.Where("message_type", MessageTypeL1SentMessage)
+			db = db.Where("message_nonce = ?", l1MessageQueueEvent.QueueIndex)
+			if err := db.Updates(updateFields).Error; err != nil {
+				return fmt.Errorf("failed to update message status as dropped, event: %+v, error: %w", l1MessageQueueEvent, err)
+			}
+		}
+	}
+	return nil
+}
+
+// UpdateBatchStatusOfL2Withdrawals updates batch status of L2 withdrawals.
+func (c *CrossMessage) UpdateBatchStatusOfL2Withdrawals(ctx context.Context, startBlockNumber, endBlockNumber, batchIndex uint64) error {
+	db := c.db.WithContext(ctx)
+	db = db.Model(&CrossMessage{})
+	db = db.Where("message_type = ?", MessageTypeL2SentMessage)
+	db = db.Where("l1_block_number >= ?", startBlockNumber)
+	db = db.Where("l1_block_number <= ?", endBlockNumber)
+	updateFields := make(map[string]interface{})
+	updateFields["batch_index"] = batchIndex
+	updateFields["rollup_status"] = RollupStatusTypeFinalized
+	if err := db.Updates(updateFields).Error; err != nil {
+		return fmt.Errorf("failed to update batch status of L2 sent messages, start: %v, end: %v, index: %v, error: %w", startBlockNumber, endBlockNumber, batchIndex, err)
+	}
+	return nil
+}
+
+// InsertOrUpdateL1Messages inserts or updates a list of L1 cross messages into the database.
+func (c *CrossMessage) InsertOrUpdateL1Messages(ctx context.Context, messages []*CrossMessage, dbTX ...*gorm.DB) error {
+	db := c.db
+	if len(dbTX) > 0 && dbTX[0] != nil {
+		db = dbTX[0]
+	}
+	db = db.WithContext(ctx)
+	db = db.Model(&CrossMessage{})
+	db = db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "message_hash"}},
+		DoUpdates: clause.AssignmentColumns([]string{"sender", "receiver", "token_type", "l1_block_number", "l1_tx_hash", "l1_token_address", "l2_token_address", "token_ids", "token_amounts", "message_type", "tx_status", "block_timestamp", "message_nonce"}),
```

Comment:
L1 update `l2_token_address` ?

Reply:
Yes. because it is the address info (to indicate the token address mapping) in the event, e.g.,
```
event DepositERC20(address indexed l1Token, address indexed l2Token, address indexed from, address to, uint256 amount, bytes data)
```